crates/database/src/order_events.rs

```diff
@@ -67,6 +67,37 @@ pub async fn insert_order_event(
         .map(|_| ())
 }
 
+/// Inserts a row into the `order_events` table only if the latest event for the
+/// corresponding order uid is not `Invalid`.
+pub async fn insert_invalid_order_event(
+    ex: &mut PgConnection,
+    event: &OrderEvent,
+) -> Result<(), sqlx::Error> {
+    const QUERY: &str = r#"
+        WITH cte AS (
+            SELECT label
+            FROM order_events
+            WHERE order_uid = $1
+            ORDER BY timestamp DESC
+            LIMIT 1
+        )
+        INSERT INTO order_events (order_uid, timestamp, label)
+        SELECT $1, $2, $3
+        WHERE NOT EXISTS (
+            SELECT 1
+            FROM cte
+            WHERE label = 'invalid'
+        )
```

Comment:
In staging, this query takes around 60ms.

Reply:
This is concerning because that doesn't bode well for prod. 😬 
Currently we discard around 520 orders from every auction because the owner does not have enough balance. If we go down this path we should probably spawn a background task for this insert as well.

Reply:
Can you execute the query like this in prod to test? Not sure if this helps, but we could add another boolean column in the table "latest" and then for every insert first update the latest record to be false. An index on this column could make lookups much faster.

Reply:
The query used in this PR is constant time since we already have a index on `(order_uid, timestamp)`.
It simply scans the index for the given `uid` backwards and because we are only interested in the last row we only do a single iteration during the scan.